### ISSUE TITLE
stream ended unexpectedly

### ISSUE THREAD

**Issue by YanshuoH** [node-apn/node-apn#617]
Hi there, 

Recently trying to use the alpha version (v3.0.0-alpha1), I've encountered the same issue posted #461 and #449 : 
```
{
  "device":"3a0b2014cd7960c037cf42da79a567e0fbeb35c988faea90944fbb9e280dd82d",
  "error":{
    "jse_shortmsg":"stream ended unexpectedly",
    "jse_info":{},
    "message":"stream ended unexpectedly"
  }
}
```

I've seen this error in two situations:
1. IDLE for a while
2. Start of the service

I downgrade `apn` to v2.*, this error disappears (rarely occurred).

**Comment by ide** [comment]
I've encountered this error several times with that same version of node-apn. From logging the device token and printing debug logs with `DEBUG=apn`, this is what I think is going on:

- When node-apn returns "stream ended unexpectedly", it's because APNs terminated the whole HTTP/2 session, and in turn terminated all of the HTTP/2 streams in the session.
- This fires the "end" event on each HTTP/2 stream before firing the "error" event on the streams. The user of node-apn doesn't see the error as a result, since the promises have already been resolved by the streams' "end" events at that point.
- I imagine that node-apn v2.x structured this logic a little differently, firing the "error" events first. This would result in error messages containing "apn write failed".
- With node-apn v2.x, I saw several "apn write failed" errors but none with v3.x. And with v3.x I began to saw several "stream ended unexpectedly" errors, which corroborates the hypothesis above. So I'm tempted to think that node-apn v3.x isn't less reliable than 2.x, and the error message has changed.
- Apple's APNs docs say they may terminate a connection by sending a GOAWAY frame with the session. However, I haven't been able to get the session's "goaway" event to fire. This is unfortunate because the GOAWAY frame would maybe contain an error body explaining what's wrong.
- Lastly, I've noticed that certain device tokens appear to trigger this error behavior. Retrying over and over is not fruitful and APNs continues to drop the connection. My guess is that the device tokens are unregistered and have continued to have been sent to APNs despite being marked as "Unregistered" before.

tl;dr: I think this error may be linked to specific device tokens and isn't necessarily node-apn's fault. The error message may have changed from node-apn v2.x to v3.x but the root cause hasn't.

If all of the failing device tokens suddenly start succeeding with node-apn v2.x and you can see the device receive the notification, that'd help indicate the issue is with v3.x.

**Comment by mikemliu** [comment]
+1 happens when provider IDLE for a time. looks to be some closed/bad connection handling issue. device tokens are good in my case. subsequent retry will succeed.

`{ VError: stream ended unexpectedly
    at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23)
    at emitNone (events.js:111:20)
    at ClientHttp2Stream.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
     jse_shortmsg: 'stream ended unexpectedly',
     jse_info: {},
     message: 'stream ended unexpectedly' }`

**Comment by mikemliu** [comment]
I can also confirm that `end` is sent before `error`. This case I think should be handled. Maybe set a delay on end where responseData is empty/status is null. In that case, we wait for the `ECONNRESET` to come thru and then can re-establish a new session and retry the request. If that `error` doesn't happen, the timer will resolve the promise with the error...

```
A  2018-08-17T00:42:49.588Z apn Request error: Error: read ECONNRESET
A  2018-08-17T00:42:49.587Z apn Request ended with status null and responseData: 
```

**Comment by ide** [comment]
I'm not sure, but the right fix might be to send a PR to Node.js upstream so the error event fires before the end event. Presumably an error (ECONNRESET) is the cause of the end event (no more data available to read). Also, on the writing side of the duplex Http2Stream, errors are emitted before close events. Someone seasoned with Node's stream semantics might be able to help make that call.

**Comment by mikemliu** [comment]
At the very least, I think the session should be destroyed when this error is encountered? I see that `this.session.destroy();` is called if a status 500 response is received. What is the intent here? If the session is to be auto-destroyed when unrecoverable state is reached, then maybe the connection reset should be included. Just depends on which session errors the library would like to handle itself and which to depend on user discretion to recycle the provider.

**Comment by ide** [comment]
The HTTP 500 status indicates an issue with Apple but a dropped stream could happen for a variety of reasons that aren't all due to a broken session, as opposed to a broken stream. That being said, I've observed that when the "stream unexpectedly ends", the session also fires an `error` event, which seems like a good reason to destroy and reconstruct the session.

With APNs's current behavior, I think node-apn should throw an async error (i.e. reject) when the stream ends due to an error. This might best be implemented with an upstream commit to Node to make it emit the "error" event before the "end" event, if that is the right semantics for Node and HTTP/2.

I also think node-apn maybe shouldn't destroy the session when a single stream experiences an error like this, and should destroy the stream when the session itself emits an error. This seems especially useful when using the JWT-based authentication, where one session may be used to send notifications with different JWTs to different apps; a session should continue to work even if one invalid JWT causes a single stream to fail; other streams using different JWTs aren't fundamentally affected regardless of how APNs's implementation behaves in this scenario.

**Comment by haroot** [comment]
yah I'm getting this error Parse Server 2.8.2


Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS VError: stream ended unexpectedly 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at emitNone (events.js:110:20) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at ClientHttp2Stream.emit (events.js:207:7) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at endReadableNT (_stream_readable.js:1056:12) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at _combinedTickCallback (internal/process/next_tick.js:138:11) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at process._tickDomainCallback [as _tickCallback] (internal/process/next_tick.js:218:9) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS  APNS error transmitting to device %s with error %s 6f7de61ef70244c1989692bcbf8b74a8e7649d41a580189e2c19434d66c92c15 { VError: stream ended unexpectedly 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at emitNone (events.js:110:20) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at ClientHttp2Stream.emit (events.js:207:7) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at endReadableNT (_stream_readable.js:1056:12) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at _combinedTickCallback (internal/process/next_tick.js:138:11) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS     at process._tickDomainCallback [as _tickCallback] (internal/process/next_tick.js:218:9) 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS   jse_shortmsg: 'stream ended unexpectedly', 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS   jse_info: {}, 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS   message: 'stream ended unexpectedly', 
Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS   stack: 'VError: stream ended unexpectedly\n    at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23)\n

solution is to update apn module?

**Comment by nareninindia** [comment]
> yah I'm getting this error Parse Server 2.8.2
> 
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS VError: stream ended unexpectedly
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at emitNone (events.js:110:20)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at ClientHttp2Stream.emit (events.js:207:7)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at endReadableNT (_stream_readable.js:1056:12)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at _combinedTickCallback (internal/process/next_tick.js:138:11)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at process._tickDomainCallback [as _tickCallback] (internal/process/next_tick.js:218:9)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS APNS error transmitting to device %s with error %s 6f7de61ef70244c1989692bcbf8b74a8e7649d41a580189e2c19434d66c92c15 { VError: stream ended unexpectedly
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at emitNone (events.js:110:20)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at ClientHttp2Stream.emit (events.js:207:7)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at endReadableNT (_stream_readable.js:1056:12)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at _combinedTickCallback (internal/process/next_tick.js:138:11)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS at process._tickDomainCallback [as _tickCallback] (internal/process/next_tick.js:218:9)
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS jse_shortmsg: 'stream ended unexpectedly',
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS jse_info: {},
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS message: 'stream ended unexpectedly',
> Jan 25 14:23:44 prod app/web.5: ERR! parse-server-push-adapter APNS stack: 'VError: stream ended unexpectedly\n at ClientHttp2Stream.request.on (/app/node_modules/apn/lib/client.js:120:23)\n
> 
> solution is to update apn module?

I was not setting the push notification configuration correctly. pem file or p12 file should be okay. But set the configuration correctly. If using pem file, the config is different. And if using p12 file, the config is different. Link to doc - https://docs.parseplatform.org/parse-server/guide/#1-prepare-apns-and-fcm-credentials

**Comment by YanshuoH** [comment]
dead.
